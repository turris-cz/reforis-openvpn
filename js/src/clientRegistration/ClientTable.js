/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";
import PropTypes from "prop-types";

import {
    formFieldsSize, SpinnerElement, useAlert, useAPIDelete, API_STATE, Button, withEither,
} from "foris";

import API_URLs from "API";
import { DownloadButton } from "foris/bootstrap/DownloadButton";

const clientShape = PropTypes.shape({
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    status: PropTypes.string.isRequired,
});

ClientTable.propTypes = {
    clients: PropTypes.arrayOf(clientShape),
    address: PropTypes.string,
};

ClientTable.defaultProps = {
    clients: [],
};

export default function ClientTable({ clients, address }) {
    if (!clients || clients.length === 0) {
        return <p className="text-muted text-center">{_("There are no clients registered.")}</p>;
    }

    const rows = clients.map((client) => (
        <ClientRow
            key={client.id}
            client={client}
            address={address}
        />
    ));

    return (
        <table className={`table table-hover ${formFieldsSize}`}>
            <thead>
                <tr>
                    <th scope="col">{_("Client name")}</th>
                    <th scope="col" aria-label={_("Actions")} />
                </tr>
            </thead>
            <tbody>
                {rows}
            </tbody>
        </table>
    );
}

ClientRow.propTypes = {
    client: clientShape.isRequired,
    address: PropTypes.string,
};

function ClientRow({ client, address }) {
    const [setAlert] = useAlert();

    const [deleteClientResponse, deleteClient] = useAPIDelete(`${API_URLs.clients}/${client.id}`);
    useEffect(() => {
        if (deleteClientResponse.state === API_STATE.ERROR) {
            setAlert(deleteClientResponse.data);
        }
    }, [deleteClientResponse, setAlert]);

    return (
        <tr>
            <td className="align-middle">{client.name}</td>
            <td className="text-center">
                <ActionsWithGeneratingAndRevoked
                    apiState={deleteClientResponse.state}
                    client={client}
                    address={address}
                    onRevoke={deleteClient}
                />
            </td>
        </tr>
    );
}

const withGenerating = withEither(
    (props) => props.apiState === API_STATE.SENDING || props.client.status === "generating",
    SpinnerElement,
);
const withRevoked = withEither(
    (props) => props.client.status === "revoked",
    () => <span>{_("Access revoked")}</span>,
);
const ActionsWithGeneratingAndRevoked = withGenerating(withRevoked(Actions));

Actions.propTypes = {
    client: clientShape.isRequired,
    address: PropTypes.string,
    onRevoke: PropTypes.func.isRequired,
    disabled: PropTypes.bool,
};

Actions.defaultProps = {
    disabled: false,
};

function Actions({
    client, address, onRevoke, disabled,
}) {
    const downloadDisabled = disabled || typeof address !== "string";
    return (
        <div className="btn-group" role="group">
            <DownloadButton
                href={`${API_URLs.clients}/${client.id}?address=${address}`}
                className={`btn-primary btn-sm ${downloadDisabled ? "disabled" : ""}`.trim()}
            >
                {_("Download")}
            </DownloadButton>
            <Button
                onClick={onRevoke}
                className="btn-danger btn-sm"
                disabled={disabled}
            >
                {_("Revoke")}
            </Button>
        </div>
    );
}
