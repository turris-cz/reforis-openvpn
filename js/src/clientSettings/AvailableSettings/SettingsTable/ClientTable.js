/*
 * Copyright (C) 2019-2022 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";

import PropTypes from "prop-types";

import ClientRow from "./ClientRow";

ClientTable.propTypes = {
    clientSettings: PropTypes.array.isRequired,
    editClientSettings: PropTypes.func.isRequired,
};

export default function ClientTable({ clientSettings, editClientSettings }) {
    if (!clientSettings || clientSettings.length === 0) {
        return (
            <p className="text-muted text-center">
                {_("There are no settings added yet.")}
            </p>
        );
    }

    return (
        <div className="table-responsive">
            <table className="table table-hover mb-0">
                <thead className="thead-light">
                    <tr>
                        <th>{_("Instances")}</th>
                        <th
                            scope="col"
                            aria-label={_("Actions")}
                            className="text-center"
                        />
                        <th
                            scope="col"
                            aria-label={_("Actions")}
                            className="text-end"
                        />
                    </tr>
                </thead>
                <tbody>
                    {clientSettings.map((client) => (
                        <ClientRow
                            key={client.id}
                            client={client}
                            editClientSettings={editClientSettings}
                        />
                    ))}
                </tbody>
            </table>
        </div>
    );
}
