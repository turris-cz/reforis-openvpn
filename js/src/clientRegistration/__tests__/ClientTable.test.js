/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import {
    render,
    fireEvent,
    wait,
    getByText,
} from "foris/testUtils/customTestRender";
import { mockSetAlert } from "foris/testUtils/alertContextMock";
import { mockJSONError } from "foris/testUtils/network";
import mockAxios from "jest-mock-axios";

import ClientTable from "../ClientTable";

describe("<ClientTable />", () => {
    const singleClient = [{ id: "A1", name: "First", status: "valid" }];

    function renderTable(clients = [], address = "") {
        const { container } = render(
            <ClientTable clients={clients} address={address} />
        );
        return container;
    }

    function getRevokeButton(container) {
        return getByText(container, "Revoke");
    }

    it("should render info when there are no clients", () => {
        const container = renderTable();
        expect(container).toMatchSnapshot();
    });

    it("should render table of clients", () => {
        const clients = [
            { id: "A1", name: "First", status: "valid" },
            { id: "A2", name: "Second", status: "generating" },
            { id: "B1", name: "Third", status: "valid" },
            { id: "B2", name: "Last", status: "revoke1d" },
        ];
        const container = renderTable(clients);
        expect(container).toMatchSnapshot();
    });

    it("should disable download on invalid address", () => {
        const clients = [{ id: "A1", name: "First", status: "valid" }];
        const container = renderTable(clients, null);
        const downloadButton = getByText(container, "Download");
        expect(
            downloadButton.parentElement.classList.contains("disabled")
        ).toBe(true);
    });

    it("should display spinner on delete request", () => {
        const container = renderTable(singleClient);
        fireEvent.click(getRevokeButton(container));
        expect(container).toMatchSnapshot();
    });

    it("should send delete request", () => {
        const container = renderTable(singleClient);
        fireEvent.click(getRevokeButton(container));
        expect(mockAxios.delete).toBeCalledWith(
            `/reforis/openvpn/api/clients/${singleClient[0].id}`,
            expect.anything()
        );
    });

    it("should handle revoke api error", async () => {
        const container = renderTable(singleClient);
        fireEvent.click(getRevokeButton(container));
        const errorMessage = "API didn't handle this well";
        mockJSONError(errorMessage);
        await wait(() => {
            expect(mockSetAlert).toHaveBeenCalledWith(errorMessage);
        });
    });
});
