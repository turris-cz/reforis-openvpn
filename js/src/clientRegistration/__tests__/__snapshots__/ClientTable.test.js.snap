// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClientTable /> should display spinner on delete request 1`] = `
<div>
  <div
    class="table-responsive"
  >
    <table
      class="table table-hover"
    >
      <thead
        class="thead-light"
      >
        <tr>
          <th
            scope="col"
          >
            Client name
          </th>
          <th
            aria-label="Actions"
            scope="col"
          />
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="align-middle"
          >
            First
          </td>
          <td
            class="text-right"
          >
            <div
              class="spinner-border spinner-border-sm"
              role="status"
            >
              <span
                class="sr-only"
              />
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`<ClientTable /> should render info when there are no clients 1`] = `
<div>
  <p
    class="text-muted text-center"
  >
    There are no clients registered.
  </p>
</div>
`;

exports[`<ClientTable /> should render table of clients 1`] = `
<div>
  <div
    class="table-responsive"
  >
    <table
      class="table table-hover"
    >
      <thead
        class="thead-light"
      >
        <tr>
          <th
            scope="col"
          >
            Client name
          </th>
          <th
            aria-label="Actions"
            scope="col"
          />
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="align-middle"
          >
            First
          </td>
          <td
            class="text-right"
          >
            <div
              class="btn-group mb-0"
              role="group"
            >
              <a
                class="btn btn-primary btn-sm"
                download=""
                href="/reforis/openvpn/api/clients/A1?address="
              >
                Download
              </a>
              <button
                class="btn btn-danger btn-sm d-inline-flex justify-content-center align-items-center"
                type="button"
              >
                <span>
                  Revoke
                </span>
              </button>
            </div>
          </td>
        </tr>
        <tr>
          <td
            class="align-middle"
          >
            Second
          </td>
          <td
            class="text-right"
          >
            <div
              class="spinner-border spinner-border-sm"
              role="status"
            >
              <span
                class="sr-only"
              />
            </div>
          </td>
        </tr>
        <tr>
          <td
            class="align-middle"
          >
            Third
          </td>
          <td
            class="text-right"
          >
            <div
              class="btn-group mb-0"
              role="group"
            >
              <a
                class="btn btn-primary btn-sm"
                download=""
                href="/reforis/openvpn/api/clients/B1?address="
              >
                Download
              </a>
              <button
                class="btn btn-danger btn-sm d-inline-flex justify-content-center align-items-center"
                type="button"
              >
                <span>
                  Revoke
                </span>
              </button>
            </div>
          </td>
        </tr>
        <tr>
          <td
            class="align-middle"
          >
            Last
          </td>
          <td
            class="text-right"
          >
            <div
              class="btn-group mb-0"
              role="group"
            >
              <a
                class="btn btn-primary btn-sm"
                download=""
                href="/reforis/openvpn/api/clients/B2?address="
              >
                Download
              </a>
              <button
                class="btn btn-danger btn-sm d-inline-flex justify-content-center align-items-center"
                type="button"
              >
                <span>
                  Revoke
                </span>
              </button>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
